File: Cycle_Sort/Cycle_Sort.js
// Cycle Sort Algorithm implementation

const cycleSort = (arr, n) => {
    // To count number of memory writes
    let writes = 0;

    // consider each index as the start of a cycle
    for (let cycle_start = 0; cycle_start <= n - 2; cycle_start++) {
        // initialize item as starting point
        let item = arr[cycle_start];

        // Correct position: All smaller elements on right side of item
        // + current_index
        let pos = cycle_start;
        for (let i = cycle_start + 1; i < n; i++)
            if (arr[i] < item) pos++;

            // If item is already in correct position
        if (pos == cycle_start) continue;

        // ignore all duplicates
        while (item == arr[pos]) pos += 1;

        // put the item to its correct position
        if (pos != cycle_start) {
            //swap
            let x = item;
            item = arr[pos];
            arr[pos] = x;
            writes++;
        }

        // Rotate rest of the cycle
        while (pos != cycle_start) {
            pos = cycle_start;

            // Find position where we put the element
            for (let i = cycle_start + 1; i < n; i++)
                if (arr[i] < item) pos += 1;

                // ignore all duplicate  elements
            while (item == arr[pos]) pos += 1;

            // put the item to it's right position
            if (item != arr[pos]) {
                //swap
                let x = item;
                item = arr[pos];
                arr[pos] = x;
                writes++;
            }
        }
    }
};

let arr = [1, 2, 8, 9, 7, 5, 5, 4, 6];
let n = arr.length;
console.log("Unsorted Array: ", arr);

cycleSort(arr, n);
console.log("Sorted Array: ", arr);

// Output:
// Unsorted Array: [1, 2, 8, 9, 7, 5, 5, 4, 6]
// Sorted Array: [1, 2, 4, 5, 5, 6, 7, 8, 9]

File: Cycle_Sort/Cycle_Sort.c
//Cycle Sort in C
#include<stdio.h>

void sort(int a[], int n)
{
    int writes = 0,start,element,pos,temp,i;
    
    // Find position where we put the element.
    for (start = 0; start <= n - 2; start++) {
        element = a[start];
        pos = start;
        for (i = start + 1; i < n; i++)
            if (a[i] < element)
                pos++;
        
        // If element is already in correct position
        if (pos == start)
            continue;
        while (element == a[pos])
            pos += 1;
        
        // put the element to it's right position
        if (pos != start) {
            temp = element;
            element = a[pos];
            a[pos] = temp;
            writes++;
        }
        
        // Rotate rest of the cycle
        while (pos != start) {
            pos = start;
            for (i = start + 1; i < n; i++)
                if (a[i] < element)
                    pos += 1;
            while (element == a[pos])
                pos += 1;
            
            // put the element to it's right position
            if (element != a[pos]) {
                temp = element;
                element = a[pos];
                a[pos] = temp;
                writes++;
            }
        }
    }
}

int main()
{
    int a[50], size;
    printf("Enter number of elements/n");
    scanf("%d",&size);
    printf("Enter array elements: \n");
    for(int i=0;i<size;i++)
    {
        printf("a[%d] :",i);
        scanf("%d",&a[i]);
    }
  
    sort(a, size);
    printf("After sort, array : \n");
    for (int i = 0; i < size; i++)
        printf("%d ",a[i]);
    return 0;
}
/*Output-
Enter number of elements/n6
Enter array elements:
a[0] :3
a[1] :9
a[2] :1
a[3] :-6
a[4] :4
a[5] :8
After sort, array :
-6 1 3 4 8 9*/
